samples/Common/cpp/CameraHelper_cppwinrt.cpp

```diff
@@ -0,0 +1,279 @@
+                   << "@" << std::to_wstring(selectedFormat.FrameRate().Numerator() / selectedFormat.FrameRate().Denominator()) << L"fps" << std::endl;
+    }
+
+    // Create FrameReader with the FrameSource that we selected in the loop above.
+    m_frameReader = m_mediaCapture.CreateFrameReaderAsync(frameSourceIterator.Current().Value()).get();
+
+    // Set up a delegate to handle the frames when they are ready
+    m_frameArrivedEventToken = m_frameReader.FrameArrived({ this, &CameraHelper::FirstFrameArrivedHandler });
+
+    // Finally start the FrameReader
+    m_frameReader.StartAsync().get();
+}
+
+//
+// Dispose of camera pipeline resources
+//
+void CameraHelper::Cleanup()
+{
+    // Revoke callback, stop FrameReader and close instances
+    if (m_frameReader != nullptr)
+    {
+        m_frameReader.FrameArrived(m_frameArrivedEventToken);
+        m_frameReader.StopAsync().get();
+        m_frameReader.Close();
+        m_frameReader = nullptr;
+    }
+    if (m_mediaCapture != nullptr)
+    {
+        m_mediaCapture.Failed(m_failureEventToken);
+        m_mediaCapture.Close();
+        m_mediaCapture = nullptr;
+    }
+}
+
+//
+// Function to handle the frame when it arrives from FrameReader
+// and send it back to registered new frame handler if it is valid
+//
+void CameraHelper::FrameArrivedHandler(MediaFrameReader FrameReader, MediaFrameArrivedEventArgs)
+{
+    MediaFrameReference mediaFrame(nullptr);
+
+    // Try to get the actual Video Frame from the FrameReader
+    mediaFrame = FrameReader.TryAcquireLatestFrame();
+    if (mediaFrame != nullptr)
+    {
+        auto vmFrame = mediaFrame.VideoMediaFrame();
+        if (vmFrame != nullptr)
+        {
+            auto videoFrame = vmFrame.GetVideoFrame();
+            m_signalFrameAvailable(videoFrame);
+        }
+        mediaFrame.Close();
+    }
+}
+
+//
+// Function to handle the first frame when it arrives from FrameReader
+// and display its format then register the main handler.
+//
+void CameraHelper::FirstFrameArrivedHandler(MediaFrameReader FrameReader, MediaFrameArrivedEventArgs args)
+{
+    auto frame = FrameReader.TryAcquireLatestFrame();
+
+    if (frame == nullptr)
+        return;
+    lock.lock();
+    if (m_firstFrameReceived != 0)
+    {
+        lock.unlock();
+        return;
+    }
+    else
+    {
+        m_firstFrameReceived++;
+        lock.unlock();
+    }
+    FrameReader.StopAsync().get();
+
+    auto selectedFormat = frame.Format();
+    std::wcout << "Receiving frame " << selectedFormat.Subtype().c_str()
+               << " : " << std::to_wstring(selectedFormat.VideoFormat().Width()) << "x" << std::to_wstring(selectedFormat.VideoFormat().Height())
+               << "@" << std::to_wstring(selectedFormat.FrameRate().Numerator() / selectedFormat.FrameRate().Denominator()) << L"fps" << std::endl;
+
+    FrameReader.FrameArrived(m_frameArrivedEventToken);
```

Comment:
Should this be locked?

Reply:
in theory, if m_firstFrameReceived != 0 you cannot reach this far, and the check for m_firstFrameReceived != 0 as well as m_firstFrameReceived++ is under lock

---
In reply to: [290460915](https://github.com/microsoft/WindowsVisionSkillsPreview/pull/28#discussion_r290460915) [](ancestors = 290460915)